crates/proof-of-sql-parser/src/posql_time/timestamp.rs

```diff
@@ -0,0 +1,226 @@
+use super::{timezone, unit::PoSQLTimeUnit};
+use crate::error::PoSQLTimestampError;
+use chrono::{offset::LocalResult, DateTime, TimeZone, Utc};
+use serde::{Deserialize, Serialize};
+
+/// Represents a fully parsed timestamp with detailed time unit and timezone information
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct PoSQLTimestamp {
+    /// The datetime representation in UTC.
+    pub timestamp: DateTime<Utc>,
+
+    /// The precision of the datetime value, e.g., seconds, milliseconds.
+    pub timeunit: PoSQLTimeUnit,
+
+    /// The timezone of the datetime, either UTC or a fixed offset from UTC.
+    pub timezone: timezone::PoSQLTimeZone,
+}
+
+impl PoSQLTimestamp {
+    /// Attempts to parse a timestamp string into an [PoSQLTimestamp] structure.
+    /// This function supports two primary formats:
+    ///
+    /// 1. **RFC 3339 Parsing**:
+    ///    - Parses the timestamp along with its timezone.
+    ///    - If parsing succeeds, it extracts the timezone offset using `dt.offset().local_minus_utc()`
+    ///      and then uses this to construct the appropriate `PoSQLTimeZone`.
+    ///
+    /// 2. **Timezone Parsing and Conversion**:
+    ///    - The `from_offset` method is used to determine whether the timezone should be represented
+    ///      as `Utc` or `FixedOffset`. This function simplifies the decision based on the offset value.
+    ///
+    /// # Examples
+    /// ```
+    /// use chrono::{DateTime, Utc};
+    /// use proof_of_sql_parser::posql_time::{timestamp::PoSQLTimestamp, timezone::PoSQLTimeZone};
+    ///
+    /// // Parsing an RFC 3339 timestamp without a timezone:
+    /// let timestamp_str = "2009-01-03T18:15:05Z";
+    /// let intermediate_timestamp = PoSQLTimestamp::try_from(timestamp_str).unwrap();
+    /// assert_eq!(intermediate_timestamp.timezone, PoSQLTimeZone::Utc);
+    ///
+    /// // Parsing an RFC 3339 timestamp with a positive timezone offset:
+    /// let timestamp_str_with_tz = "2009-01-03T18:15:05+03:00";
+    /// let intermediate_timestamp = PoSQLTimestamp::try_from(timestamp_str_with_tz).unwrap();
+    /// assert_eq!(intermediate_timestamp.timezone, PoSQLTimeZone::FixedOffset(10800)); // 3 hours in seconds
+    /// ```
+    pub fn try_from(timestamp_str: &str) -> Result<Self, PoSQLTimestampError> {
+        let dt = DateTime::parse_from_rfc3339(timestamp_str)
+            .map_err(|e| PoSQLTimestampError::ParsingError(e.to_string()))?;
+
+        let offset_seconds = dt.offset().local_minus_utc();
+        let timezone = timezone::PoSQLTimeZone::from_offset(offset_seconds);
+        let nanoseconds = dt.timestamp_subsec_nanos();
+        let timeunit = if nanoseconds % 1_000 != 0 {
+            PoSQLTimeUnit::Nanosecond
+        } else if nanoseconds % 1_000_000 != 0 {
+            PoSQLTimeUnit::Microsecond
+        } else if nanoseconds % 1_000_000_000 != 0 {
+            PoSQLTimeUnit::Millisecond
+        } else {
+            PoSQLTimeUnit::Second
+        };
```

Comment:
Try to add some tests for this

Reply:
tests for this are already present in the [unit](https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/84c6717769e50ea79bd20fbb3e71abe6af34e443/crates/proof-of-sql-parser/src/posql_time/unit.rs#L55-L85) module

Reply:
tests for the ```Second``` case are found [here](https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/84c6717769e50ea79bd20fbb3e71abe6af34e443/crates/proof-of-sql-parser/src/posql_time/timestamp.rs#L169-L176)